### PR TITLE
[3109] Refactor AttachmentViewFactory

### PR DESCRIPTION
**Pull request by bychkovdmitry**
### 🎯 Goal

Closes #3116 

As we now have separate ViewHolders to handle image, file and giphy attachments, `TextAndAttachmentsViewHolder` doesn't pull its weight. We need to make it explicit that the ViewHolder is now only responsible for rendering messages with custom attachments. That's why we need to rename it to `CustomAttachmentsViewHolder`. Also, we need to refactor `AttachmentViewFactory` to make it easy to register custom attachment types.

### 🎨 UI Changes

The UI with the patch from the **testing** section.

| Message List | Message Overlay |
| --- | --- |
| photo_2022-02-23 23 57 18 | photo_2022-02-23 23 57 22 |

### 🧪 Testing

Apply the code from the snippet below see how the support for a custom attachment can be added:

<details>
  <summary>Patch file to apply these changes for testing</summary>

```
Index: stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt
--- a/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt	(revision 88caa25d0f0d703e3a342e43991560be205563fe)
+++ b/stream-chat-android-ui-components-sample/src/main/kotlin/io/getstream/chat/ui/sample/application/App.kt	(date 1645649608077)
@@ -2,12 +2,22 @@
 
 import android.app.Application
 import android.os.Build
+import android.view.ViewGroup
+import android.widget.TextView
+import androidx.core.view.setPadding
 import coil.Coil
 import coil.ImageLoader
 import coil.decode.GifDecoder
 import coil.decode.ImageDecoderDecoder
+import io.getstream.chat.android.client.logger.ChatLogger
+import io.getstream.chat.android.client.models.Message
 import io.getstream.chat.android.client.utils.internal.toggle.ToggleService
 import io.getstream.chat.android.core.internal.InternalStreamChatApi
+import io.getstream.chat.android.ui.ChatUI
+import io.getstream.chat.android.ui.message.list.adapter.MessageListListenerContainer
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.attachment.AttachmentFactories
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.attachment.AttachmentFactory
+import io.getstream.chat.android.ui.message.list.adapter.viewholder.attachment.AttachmentViewHolder
 import io.getstream.chat.ui.sample.BuildConfig
 import io.getstream.chat.ui.sample.data.user.SampleUser
 import io.getstream.chat.ui.sample.data.user.UserRepository
@@ -34,6 +44,8 @@
         )
         ApplicationConfigurator.configureApp(this)
         initializeToggleService()
+
+        ChatUI.attachmentFactories = AttachmentFactories(listOf(CustomAttachmentFactory()))
     }
 
     private fun getApiKey(): String {
@@ -55,3 +67,56 @@
             private set
     }
 }
+
+class CustomAttachmentFactory : AttachmentFactory {
+    override fun canHandle(message: Message): Boolean {
+        return message.attachments.isNotEmpty()
+    }
+
+    override fun createViewHolder(
+        message: Message,
+        listeners: MessageListListenerContainer?,
+        parent: ViewGroup,
+    ): AttachmentViewHolder {
+        return TextView(parent.context)
+            .apply { setPadding(16) }
+            .let { CustomAttachmentViewHolder(it, listeners) }
+    }
+
+    class CustomAttachmentViewHolder(
+        private val textView: TextView,
+        listeners: MessageListListenerContainer?,
+    ) : AttachmentViewHolder(textView) {
+
+        private lateinit var message: Message
+
+        init {
+            textView.setOnClickListener {
+                listeners?.attachmentClickListener?.onAttachmentClick(message, message.attachments.first())
+            }
+            textView.setOnLongClickListener {
+                listeners?.messageLongClickListener?.onMessageLongClick(message)
+                true
+            }
+        }
+
+        override fun onBindViewHolder(message: Message) {
+            ChatLogger.get("ABC").logE("ABC bind")
+            this.message = message
+
+            textView.text = "Custom attachment of type: " + message.attachments.first().type
+        }
+
+        override fun onUnbindViewHolder() {
+            ChatLogger.get("ABC").logE("ABC unbind")
+        }
+
+        override fun onViewAttachedToWindow() {
+            ChatLogger.get("ABC").logE("ABC attach")
+        }
+
+        override fun onViewDetachedFromWindow() {
+            ChatLogger.get("ABC").logE("ABC detach")
+        }
+    }
+}

```

</details>

#### General
- [x] I have signed the [Stream CLA](https://docs.google.com/forms/d/e/1FAIpQLScFKsKkAJI7mhCr7K9rEIOpqIDThrWxuvxnwUq2XkHyG154vQ/viewform) (required)
- [x] Assigned a person / code owner group (required)
- [x] Thread with the PR link started in a respective Slack channel (#android-chat-core or #android-chat-ui) (required)
- [x] PR targets the `develop` branch
- [x] PR is linked to the GitHub issue it resolves

#### Code & documentation
- [x] Changelog is updated with client-facing changes
- [ ] New code is covered by unit tests
- [x] Comparison screenshots added for visual changes
- [x] Affected documentation updated (KDocs, docusaurus, tutorial)

### ☑️Reviewer Checklist
- [ ] UI Components sample runs & works
- [ ] Compose sample runs & works
- [ ] UI Changes correct (before & after images)
- [ ] Bugs validated (bugfixes)
- [ ] New feature tested and works
- [ ] Release notes and docs clearly describe changes
- [ ] All code we touched has new or updated KDocs

### CODE REVIEW

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/attachment/AttachmentViewHolder.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+import android.view.View
+import io.getstream.chat.android.client.models.Message
+
+/**
+ * An inner ViewHolder for custom attachments within message ViewHolder.
+ * Follows the lifecycle of the outer message ViewHolder.
+ *
+ * @param itemView The view that this ViewHolder controls.
+ */
+public abstract class AttachmentViewHolder(public val itemView: View) {
```
Comment (MarinTolic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/attachment/AttachmentViewHolder.kt:13]: Calling this a `ViewHolder` might be slightly confusing for users who see it and expect a `RecyclerView.ViewHolder`.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/attachment/AttachmentFactories.kt [hunk 2]
```diff
@@ -0,0 +1,40 @@
+package io.getstream.chat.android.ui.message.list.adapter.viewholder.attachment
+
+import android.view.ViewGroup
+import io.getstream.chat.android.client.models.Message
+import io.getstream.chat.android.ui.message.list.adapter.MessageListListenerContainer
+
+/**
+ * A manager for registered custom attachment factories.
+ */
+public class AttachmentFactories(
```
Comment (filbabic) [inline stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/attachment/AttachmentFactories.kt:10]: Maybe `AttachmentFactory(Factories)Manager`?

**Review by filbabic — APPROVED** [review]
Left a couple of comments that you can take a look at but I don't have any outstanding issues with this, it seems like exactly what we're going for!

### ☑️Reviewer Checklist
- [x] UI Components sample runs & works
- [x] Compose sample runs & works
- [x] UI Changes correct (before & after images)
- [x] New feature tested and works
- [x] Release notes and docs clearly describe changes
- [x] All code we touched has new or updated KDocs

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/attachment/AttachmentFactories.kt (hunk 2 above)
Comment (bychkovdmitry) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/attachment/AttachmentFactories.kt]: Yes, agree, no need to make this the "Compose" way. Renamed to `AttachmentFactoryManager`.

File: stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/internal/CustomAttachmentsViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    /**
+     * Updates the custom attachments section of the message.
+     */
+    private fun bindCustomAttachments(data: MessageListItem.MessageItem) {
+        this.attachmentViewHolder = attachmentFactories.createViewHolder(data.message, listeners, binding.root)
+            .also { attachmentViewHolder ->
+                attachmentViewHolder.onBindViewHolder(data.message)
+                binding.attachmentsContainer.removeAllViews()
+                binding.attachmentsContainer.addView(attachmentViewHolder.itemView)
```
[the remark this answers is not in the document]
Comment (bychkovdmitry) [reply stream-chat-android-ui-components/src/main/kotlin/io/getstream/chat/android/ui/message/list/adapter/viewholder/internal/CustomAttachmentsViewHolder.kt:88]: I think we need lifecycle methods. I will merge as is for now and we can discuss a better naming for this inner ViewHolder delegate tomorrow.

**Pull request merged**